### DISCUSSION TITLE
Dynamically inserting fields in a form and capturing them in a submission?

### DISCUSSION THREAD

**Discussion by chrismlusk** [verbb/formie#1364] [Q&A]
This might be an odd one, but I'm curious if there's a way for a submission to include fields that are not part of the Formie form.

I have a client who regularly micromanages the tracking parameters in URLs for their paid campaigns. Because they have many forms, I want to avoid having to edit every form whenever they want to add/remove tracking parameters that should be included in the form submission.

Currently, they have a Global field where they list the query parameters they want to track. Our base template grabs those and stores them on the `window`. Then, whenever a form loads (I'm using Freeform currently), I grab any query parameters from the URL that match these "tracked" params, I create a new hidden `input` with that param's value, and I add it to the `form` element. Of course Freeform is unaware of these — so they don't show up in submissions or anything — but because I'm forwarding the submission to a third party (per my question in #1361), I just need them included in the payload.

This is where I'm blocked with Formie. Any ideas how I could work around this?

**Comment by engram-design** [comment]
Submissions can only hold fields that are Formie field, not Craft fields, or other data. In order to record them against the submission, they'd need to be stored on it somehow. Compared to Freeform, Formie does work a little differently where this isn't possible like it is with Freeform, because integrations fire through the queue. With Freeform that data is there in POST data, but that's no longer the case with Formie and the queue.

It might be a neat idea to be able to store arbitrary extra data on a submission, but it still wouldn't be overly useful, because you would have to define what to do with that data.

You can certainly also modify the payload that the integration sends via events if you have some PHP and module know-how. That way you can keep it completely out of the front-end altogether.

I suppose I would suggest to create a hidden input on every form to store this data. You can use `populateFormValues()` to populate that hidden field from globals. It does just require you to create those hidden fields on the form.

**Reply by chrismlusk** [reply]
I think you are right that storing arbitrary data on submissions wouldn't be useful in most circumstances, but I do think it's a nice idea that could be beneficial for developers doing custom work.  

In the meantime, I like your suggestion. Thanks!

**Reply by engram-design** [reply]
Yep, I'll give it some further thought! I did also have a thought, you could use the `snapshot` functionality for this, which handles arbitrary data for you to access later. In practice, this stores modifications to fields like Dropdown fields, where you might change the options of them at runtime. These are stored against the submission and then pulled in. Lots of people like to use dynamic options for Dropdown (and other) fields, despite it going against their concept.

You can use:

```twig
{% do form.setSnapshotData({ some: 'data' }) %}
```

and if you wanted to access that in PHP in an event, custom integration or otherwise, 

```php
$form->getSnapshotData();
```

**Reply by chrismlusk** [reply]
Nice! The `snapshot` works great for this. Thanks so much for sharing.

One quick correction for anyone who ends up here in the future, the `setSnapshotData` method requires two arguments:

```php
public function setSnapshotData($key, $data): void
```
